segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            return FileSystemSegmentHandle.getWriteHandle(streamSegmentName);
+        } else {
+                return openRead(streamSegmentName).get();
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) {
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+            if (Files.size(path) < offset) {
+                log.warn("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
+                        handle.getSegmentName(), offset);
+                throw new IllegalArgumentException( "Reading beyond the current size of segment");
+            }
+
+        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) {
+            int bytesRead = channel.read(ByteBuffer.wrap(buffer, bufferOffset, length), offset);
+            return bytesRead;
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private SegmentProperties syncGetStreamSegmentInfo(String streamSegmentName) {
+            PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getFilesystemRoot(), streamSegmentName),
+                    PosixFileAttributes.class);
+            StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                    !(attrs.permissions().contains(OWNER_WRITE)), false,
+                    new ImmutableDate(attrs.creationTime().toMillis()));
+
+            return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        boolean exists = Files.exists(Paths.get(config.getFilesystemRoot(), streamSegmentName));
+        return exists;
+    }
+
+    @SneakyThrows
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        log.info("Creating Segment {}", streamSegmentName);
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, null).get();
```

Comment:
Using the sync method instead. I think that is the right thing to do rather than creating a dummy object.